asset-transfer-ledger-queries/chaincode-javascript/lib/asset_transfer_ledger_chaincode.js

```diff
@@ -0,0 +1,425 @@
+
+  // getQueryResultForQueryString executes the passed in query string.
+  // Result set is built and returned as a byte array containing the JSON results.
+  async getQueryResultForQueryString(ctx, queryString) {
+    
+    let resultsIterator = await ctx.stub.getQueryResult(queryString);
+    let results = await this.getAllResults(resultsIterator, false);
+
+    return JSON.stringify(results);
+  }
+
+  // Example: Pagination with Range Query
+  // GetAssetsByRangeWithPagination performs a range query based on the start & end key,
+  // page size and a bookmark.
+  // The number of fetched records will be equal to or lesser than the page size.
+  // Paginated range queries are only valid for read only transactions.
+  async getAssetsByRangeWithPagination(ctx, startKey, endKey, pageSize, bookmark) {
+    
+    const { iterator, metadata } = await ctx.stub.getStateByRangeWithPagination(startKey, endKey, pageSize, bookmark);
+    const results = await this.getAllResults(iterator, false);
+
+    results.ResponseMetadata = {
+      RecordsCount: metadata.fetched_records_count,
+      Bookmark: metadata.bookmark,
+    };
+    return JSON.stringify(results);
+  }
+
+  // Example: Pagination with Ad hoc Rich Query
+  // QueryAssetsWithPagination uses a query string, page size and a bookmark to perform a query
+  // for assets. Query string matching state database syntax is passed in and executed as is.
+  // The number of fetched records would be equal to or lesser than the specified page size.
+  // Supports ad hoc queries that can be defined at runtime by the client.
+  // If this is not desired, follow the QueryAssetsForOwner example for parameterized queries.
+  // Only available on state databases that support rich query (e.g. CouchDB)
+  // Paginated queries are only valid for read only transactions.
+  async queryAssetsWithPagination(ctx, queryString, pageSize, bookmark) {
+
+    const { iterator, metadata } = await ctx.stub.getQueryResultWithPagination(queryString, pageSize, bookmark);
+    const results = await this.getAllResults(iterator, false);
+
+    results.ResponseMetadata = {
+      RecordsCount: metadata.fetched_records_count,
+      Bookmark: metadata.bookmark,
+    };
+
+    return JSON.stringify(results);
+  }
+
+  // GetAssetHistory returns the chain of custody for an asset since issuance.
+  async getAssetHistory(ctx, assetName) {
+
+    let resultsIterator = await ctx.stub.getHistoryForKey(assetName);
+    let results = await this.getAllResults(resultsIterator, true);
+
+    return JSON.stringify(results);
+  }
+
+  // AssetExists returns true when asset with given ID exists in world state
+  async assetExists(ctx, assetName) {
+    // ==== Check if asset already exists ====
+    let assetState = await ctx.stub.getState(assetName);
+    if ( !assetState || assetState.length === 0 ) {
+      return false;
+    }
+    return true
+  }
+
+  // getAllAssets returns all assets found in the world state.
+  async getAllAssets(ctx) {
```

Comment:
Remove this function. Clients can do a range query and pass empty strings instead.